File: src/main/java/io/wdsj/asw/util/context/ChatContext.java
package io.wdsj.asw.util.context;

import io.wdsj.asw.setting.PluginSettings;
import org.bukkit.entity.Player;

import java.util.Deque;
import java.util.LinkedList;
import java.util.concurrent.ConcurrentHashMap;

import static io.wdsj.asw.AdvancedSensitiveWords.settingsManager;

public class ChatContext {
    private static final ConcurrentHashMap<Player, Deque<String>> chatHistory = new ConcurrentHashMap<>();
    /**
     * Add player message to history
     */
    public static void addMessage(Player player, String message) {
        chatHistory.computeIfAbsent(player, k -> new LinkedList<>());
        Deque<String> history = chatHistory.get(player);
        while (history.size() >= settingsManager.getProperty(PluginSettings.CHAT_CONTEXT_MAX_SIZE)) {
            history.poll();
        }
        history.offer(message.trim());
    }

    public static Deque<String> getHistory(Player player) {
        return chatHistory.getOrDefault(player, new LinkedList<>());
    }

    public static void clearPlayerContext(Player player) {
        if (chatHistory.get(player) == null) return;
        chatHistory.remove(player);
    }

    public static void removePlayerContext(Player player) {
        Deque<String> history = chatHistory.get(player);
        if (history != null) history.pollLast();
    }
    public static void forceClearContext() {
        chatHistory.clear();
    }

    private ChatContext() {}
}

File: paper/src/main/kotlin/io/wdsj/asw/bukkit/util/context/ChatContext.kt
package io.wdsj.asw.bukkit.util.context

import io.wdsj.asw.bukkit.AdvancedSensitiveWords.setting
import io.wdsj.asw.bukkit.setting.PluginSettings
import io.wdsj.asw.common.datatype.TimedString
import org.bukkit.entity.Player

object ChatContext {
    private val chatHistory = ContextHistory<TimedString>()

    /**
     * Add player message to history
     */
    fun addMessage(player: Player, message: String) {
        message.trim().takeIf { it.isNotEmpty() }?.let {
            chatHistory.add(player.uniqueId, contextCapacity(), TimedString.of(it))
        }
    }

    fun getHistory(player: Player): List<String> {
        return chatHistory.snapshot(
            player.uniqueId,
            contextCapacity(),
            setting(PluginSettings.CHAT_CONTEXT_TIME_LIMIT) * 1_000L,
        ) { it.time }.map { it.string }
    }

    fun clearPlayerContext(player: Player) {
        chatHistory.clear(player.uniqueId)
    }

    fun pollPlayerContext(player: Player) {
        chatHistory.removeLast(player.uniqueId)
    }

    @JvmStatic
    fun forceClearContext() {
        chatHistory.clearAll()
    }

    private fun contextCapacity(): Int = setting(PluginSettings.CHAT_CONTEXT_MAX_SIZE)
}
